miden-lib/asm/miden/kernels/tx/account.masm

```diff
@@ -469,39 +475,79 @@ export.set_map_item.3
     # => [OLD_MAP_ROOT, OLD_VALUE, ...]
 end
 
-#! Verifies that the procedure root is part of the account code Merkle tree. Panics if the
-#! procedure root is not part of the account code Merkle tree.
+#! Returns the procedure information
+#!
+#! Stack: [index, ...]
+#! Output: [PROC_ROOT, storage_offset, ...]
+#!
+#! - PROC_ROOT is the hash of the procedure.
+#! - storage_offset is the procedure storage offset.
+#!
+#! Panics if
+#! - index is out of bounds
+export.get_procedure_info
+  # TODO: Fix VM caller == [0,0,0,0] bug and remove this check
+  # check if index == 255
+  dup push.255 neq
+  # => [is_255, index]
+
+  # load procedure information from memory
+  if.true
+    # check that index < number of procedures contained in the account code
+    dup exec.memory::get_num_account_procedures lt assert.err=ERR_PROC_INDEX_OUT_OF_BOUNDS
+    # => [index]
+
+    # get procedure section ptr
+    push.2 mul exec.memory::get_account_procedures_section_offset add dup push.1 add
+    # => [proc_ptr, offset_ptr]
+
+    # load procedure information from memory
+    mem_load swap padw movup.4 mem_loadw
+    # => [PROC_ROOT, storage_offset]
+  end
+ end
+
+#! Verifies that the procedure root is part of the account code
 #!
 #! Stack: [PROC_ROOT]
-#! Output: [PROC_ROOT]
+#! Output: [storage_offset]
 #!
 #! - PROC_ROOT is the hash of the procedure to authenticate.
-export.authenticate_procedure.1
-    # load the account code root onto the stack
-    exec.memory::get_acct_code_root swapw
-    # => [PROC_ROOT, CODE_ROOT]
-
-    # load the index of the procedure root onto the advice stack, and move it to the operand stack
-    emit.ACCOUNT_PUSH_PROCEDURE_INDEX_EVENT adv_push.1 movdn.4
-    # => [PROC_ROOT, index, CODE_ROOT]
-
-    # push the depth of the code Merkle tree onto the stack
-    push.ACCOUNT_CODE_TREE_DEPTH movdn.4
-    # => [PROC_ROOT, depth, index, CODE_ROOT]
-
-    # verify the procedure exists in the account code Merkle tree
-    mtree_verify
-    # => [PROC_ROOT, depth, index, CODE_ROOT]
-
-    # drop accessory variables
-    movup.4 drop movup.4 drop swapw dropw
-    # => [PROC_ROOT]
+#!
+#! Panics if
+#! - procedure root is not part of the account code.
+export.authenticate_procedure
+    # load procedure index
+    emit.ACCOUNT_PUSH_PROCEDURE_INDEX_EVENT adv_push.1
+    # => [index, PROC_ROOT]
+
+    # get procedure info (PROC_ELEMENTS, storage_offset) from memory stored at index
+    exec.get_procedure_info
+    # In production:
+    # => [PROC_ELEMENTS, storage_offset, PROC_ROOT]
+    # During testing:
+    # => [255, PROC_ROOT]
+
+    # TODO: Fix VM caller == [0,0,0,0] bug and remove this check
+    # check if get_procedure_info returned 255
+    dup push.255 neq
+    # => [is_255, index, PROC_ROOT]
```

Comment:
Same comment as above: we should try to fix this in the next PR.

Reply:
agreed.